### PR TITLE
Enable "rabbitmq_prometheus" in RabbitMQ 3.8 by default

### LINKED ISSUE #419: Consider providing a Prometheus variant
This plugin has a variant/tag where it pre-enables the management plugin. Should we add another one that enables [RabbitMQ Prometheus plugin](https://www.rabbitmq.com/prometheus.html) instead? Or would it be too much? (we definitely don't want to have a tag per plugin but these two monitoring-related ones are essential)

Enabling both makes sense in some cases and not so much in others. We can assume that Prometheus users would not want to use `rabbitmq-management` for monitoring.

### PR DESCRIPTION
**Pull request by tianon**
Closes #419

### CODE REVIEW

**Comment by tianon** [comment]
Arg, we're butting up against:

https://github.com/docker-library/rabbitmq/blob/e5da52083f1942b8a0b419045ee8a9c37958ce0e/docker-entrypoint.sh#L225-L228

I really don't like this script...  :facepalm:

**Comment by gerhard** [comment]
> I really don't like this script... 🤦

I couldn't agree more.

I would definitely advocate for its removal and reduce the sprawl of environment variables by requiring users to configure most aspects of RabbitMQ using the sysctl config format. This has built-in validation and many improvements over env vars or the classic config: https://www.rabbitmq.com/configure.html#config-file

A relevant improvement that shipped in <kbd>3.8.4</kbd> is improved support for multiple sysctl config files: https://github.com/rabbitmq/rabbitmq-server/pull/2258 & https://github.com/rabbitmq/rabbitmq-server/pull/2277 . This allows users to specify multiple config files in sysctl format and not mutate a single file.

There has never been a better time to put an end to the `docker-entrypoint,.sh` madness!

---

As extra context, this is exactly what we are doing in the commercial version of RabbitMQ, which ships as an OCI.

File: 3.7-rc/alpine/docker-entrypoint.sh
Comment (gerhard) [inline 3.7-rc/alpine/docker-entrypoint.sh:410]: I am not sure how much of the <kbd>rabbitmq_prometheus</kbd> config should be generate here, but `prometheus.return_per_object_metrics = true` is especially useful.

FWIW, this is a more compact version of all supported config properties: https://github.com/rabbitmq/rabbitmq-prometheus#configuration

**Comment by tianon** [comment]
The only reason I'm touching `docker-entrypoint.sh` here is for:

> SSL will need to be configured for rabbitmq-prometheus on top of rabbitmq-management

... and that's also why it only supports the SSL-related flags that are supported for the server and for the management plugin.

If you think we don't need that (and that if users configure the server to use SSL, we can leave it up to them to configure SSL for prometheus), we can drop all but the change to `disable_metrics_collector`.

**Comment by tianon** [comment]
We could also skip the `docker-entrypoint.sh` changes here and add a deprecation warning to `docker-entrypoint.sh` separately so that users who would be affected by the removal of it know that we intend to remove it / delete most of it at some point.

**Comment by tianon** [comment]
Also, I guess moving `disable_metrics_collector` to use a `conf.d` file instead makes it both easier to add/remove and makes this script unaware of it, so that's probably a better solution to what I pushed in https://github.com/docker-library/rabbitmq/compare/6bda400327dce532bacff769ea8b908125c7515d..e2cf7d1bbc5825230b5bcc79725816fc38bb55cf.

**Comment by nesc58** [comment]
How can I configure `return_per_object_metrics configurable`? For me it's a need to configure it.

Thank you for initiating this PR 👍

**Comment by tianon** [comment]
@nesc58 I honestly find joy in the fact that you're asking, because that very variety of question is why I hesitated even adding the configuration bits I did here :smile: -- the end goal is to not provide *any* custom configuration behavior anymore, so the direct answer to your question is that it should be put in `/etc/rabbitmq/rabbitmq.conf` or `/etc/rabbitmq/conf.d/foo.conf` (as per standard upstream RabbitMQ configuration) :+1:

File: 3.7-rc/alpine/docker-entrypoint.sh
Comment (tianon) [reply 3.7-rc/alpine/docker-entrypoint.sh:413]: If the goal is to remove this script and deprecate the logic in it, I think we should go the other direction and remove all `prometheus`-related support from this script/PR (which I'm working on now).

This means if folks want `rabbitmq_prometheus` to use TLS, they'll have to configure it to do so, but I think that's probably a good conversation starter for them to do the other TLS configuration themselves directly too (instead of using our helper environment variables) because that's the end goal anyhow.

Thoughts?

Comment (tianon) [reply 3.7-rc/alpine/docker-entrypoint.sh:413]: More concretely, here's what I'm proposing this PR be simplified to (showing only the template changes):

<details>
<summary>Diff:</summary>

```diff
diff --git a/Dockerfile-alpine.template b/Dockerfile-alpine.template
index b255bd4..f29ac98 100644
--- a/Dockerfile-alpine.template
+++ b/Dockerfile-alpine.template
@@ -174,9 +174,9 @@ ENV RABBITMQ_DATA_DIR=/var/lib/rabbitmq
 RUN set -eux; \
 	addgroup -g 101 -S rabbitmq; \
 	adduser -u 100 -S -h "$RABBITMQ_DATA_DIR" -G rabbitmq rabbitmq; \
-	mkdir -p "$RABBITMQ_DATA_DIR" /etc/rabbitmq /tmp/rabbitmq-ssl /var/log/rabbitmq; \
-	chown -fR rabbitmq:rabbitmq "$RABBITMQ_DATA_DIR" /etc/rabbitmq /tmp/rabbitmq-ssl /var/log/rabbitmq; \
-	chmod 777 "$RABBITMQ_DATA_DIR" /etc/rabbitmq /tmp/rabbitmq-ssl /var/log/rabbitmq; \
+	mkdir -p "$RABBITMQ_DATA_DIR" /etc/rabbitmq /etc/rabbitmq/conf.d /tmp/rabbitmq-ssl /var/log/rabbitmq; \
+	chown -fR rabbitmq:rabbitmq "$RABBITMQ_DATA_DIR" /etc/rabbitmq /etc/rabbitmq/conf.d /tmp/rabbitmq-ssl /var/log/rabbitmq; \
+	chmod 777 "$RABBITMQ_DATA_DIR" /etc/rabbitmq /etc/rabbitmq/conf.d /tmp/rabbitmq-ssl /var/log/rabbitmq; \
 	ln -sf "$RABBITMQ_DATA_DIR/.erlang.cookie" /root/.erlang.cookie
 
 # Use the latest stable RabbitMQ release (https://www.rabbitmq.com/download.html)
@@ -229,6 +229,8 @@ RUN set -eux; \
 	su-exec rabbitmq rabbitmq-plugins list; \
 # no stale cookies
 	rm "$RABBITMQ_DATA_DIR/.erlang.cookie"
+# Enable Prometheus-style metrics by default (https://github.com/docker-library/rabbitmq/issues/419)
+RUN set -eux; rabbitmq-plugins enable --offline rabbitmq_prometheus; echo 'management_agent.disable_metrics_collector = true' > /etc/rabbitmq/conf.d/management_agent.disable_metrics_collector.conf
 
 # Added for backwards compatibility - users can simply COPY custom plugins to /plugins
 RUN ln -sf /opt/rabbitmq/plugins /plugins
@@ -246,5 +248,5 @@ ENV LANG=C.UTF-8 LANGUAGE=C.UTF-8 LC_ALL=C.UTF-8
 COPY docker-entrypoint.sh /usr/local/bin/
 ENTRYPOINT ["docker-entrypoint.sh"]
 
-EXPOSE 4369 5671 5672 25672
+EXPOSE 4369 5671 5672 15691 15692 25672
 CMD ["rabbitmq-server"]
diff --git a/Dockerfile-management.template b/Dockerfile-management.template
index 0177890..febf166 100644
--- a/Dockerfile-management.template
+++ b/Dockerfile-management.template
@@ -1,6 +1,8 @@
 FROM %%FROM%%
 
 RUN rabbitmq-plugins enable --offline rabbitmq_management
+# make sure the metrics collector is re-enabled (disabled in the base image for Prometheus-style metrics by default)
+RUN rm -f /etc/rabbitmq/conf.d/management_agent.disable_metrics_collector.conf
 
 # extract "rabbitmqadmin" from inside the "rabbitmq_management-X.Y.Z.ez" plugin zipfile
 # see https://github.com/docker-library/rabbitmq/issues/207
diff --git a/Dockerfile-ubuntu.template b/Dockerfile-ubuntu.template
index e939ee7..84de22a 100644
--- a/Dockerfile-ubuntu.template
+++ b/Dockerfile-ubuntu.template
@@ -185,9 +185,9 @@ ENV RABBITMQ_DATA_DIR=/var/lib/rabbitmq
 RUN set -eux; \
 	groupadd --gid 999 --system rabbitmq; \
 	useradd --uid 999 --system --home-dir "$RABBITMQ_DATA_DIR" --gid rabbitmq rabbitmq; \
-	mkdir -p "$RABBITMQ_DATA_DIR" /etc/rabbitmq /tmp/rabbitmq-ssl /var/log/rabbitmq; \
-	chown -fR rabbitmq:rabbitmq "$RABBITMQ_DATA_DIR" /etc/rabbitmq /tmp/rabbitmq-ssl /var/log/rabbitmq; \
-	chmod 777 "$RABBITMQ_DATA_DIR" /etc/rabbitmq /tmp/rabbitmq-ssl /var/log/rabbitmq; \
+	mkdir -p "$RABBITMQ_DATA_DIR" /etc/rabbitmq /etc/rabbitmq/conf.d /tmp/rabbitmq-ssl /var/log/rabbitmq; \
+	chown -fR rabbitmq:rabbitmq "$RABBITMQ_DATA_DIR" /etc/rabbitmq /etc/rabbitmq/conf.d /tmp/rabbitmq-ssl /var/log/rabbitmq; \
+	chmod 777 "$RABBITMQ_DATA_DIR" /etc/rabbitmq /etc/rabbitmq/conf.d /tmp/rabbitmq-ssl /var/log/rabbitmq; \
 	ln -sf "$RABBITMQ_DATA_DIR/.erlang.cookie" /root/.erlang.cookie
 
 # Use the latest stable RabbitMQ release (https://www.rabbitmq.com/download.html)
@@ -247,6 +247,8 @@ RUN set -eux; \
 	gosu rabbitmq rabbitmq-plugins list; \
 # no stale cookies
 	rm "$RABBITMQ_DATA_DIR/.erlang.cookie"
+# Enable Prometheus-style metrics by default (https://github.com/docker-library/rabbitmq/issues/419)
+RUN set -eux; rabbitmq-plugins enable --offline rabbitmq_prometheus; echo 'management_agent.disable_metrics_collector = true' > /etc/rabbitmq/conf.d/management_agent.disable_metrics_collector.conf
 
 # Added for backwards compatibility - users can simply COPY custom plugins to /plugins
 RUN ln -sf /opt/rabbitmq/plugins /plugins
@@ -264,5 +266,5 @@ ENV LANG=C.UTF-8 LANGUAGE=C.UTF-8 LC_ALL=C.UTF-8
 COPY docker-entrypoint.sh /usr/local/bin/
 ENTRYPOINT ["docker-entrypoint.sh"]
 
-EXPOSE 4369 5671 5672 25672
+EXPOSE 4369 5671 5672 15691 15692 25672
 CMD ["rabbitmq-server"]
diff --git a/update.sh b/update.sh
index c3d66e7..365485f 100755
--- a/update.sh
+++ b/update.sh
@@ -145,5 +145,17 @@ for version in "${versions[@]}"; do
 			-e "s!%%INSTALL_PYTHON%%!$installPython!g" \
 			Dockerfile-management.template \
 			> "$version/$variant/management/Dockerfile"
+
+		# rabbitmq_prometheus is only included in 3.8+
+		# same with "/etc/rabbitmq/conf.d" support (https://github.com/rabbitmq/rabbitmq-server/pull/2258, https://github.com/rabbitmq/rabbitmq-server/pull/2277)
+		if [ "$rcVersion" = '3.7' ]; then
+			sed -ri \
+				-e '/[Pp]rometheus/d' \
+				-e '/metrics_collector/d' \
+				-e 's/ 15691 15692 / /g' \
+				-e 's! /etc/rabbitmq/conf.d ! !g' \
+				"$version/$variant/Dockerfile" \
+				"$version/$variant/management/Dockerfile"
+		fi
 	done
 done
```

</details>

Separately, I'd like to also make a PR to start showing a large warning to users for whom we're currently generating a configuration file so that they have some way to know that we intend to remove most of this script's functionality at some later date.

Comment (gerhard) [reply 3.7-rc/alpine/docker-entrypoint.sh:413]: > If the goal is to remove this script and deprecate the logic in it, I think we should go the other direction and remove all prometheus-related support from this script/PR (which I'm working on now).

Yes, agreed.

> More concretely, here's what I'm proposing this PR be simplified to

Looking good!

The only issue that I can foresee is if someone uses the non-management image and enables the management plugin. In that case, they will need to manually delete `/etc/rabbitmq/conf.d/management_agent.disable_metrics_collector.conf` in order for the UI to look as expected - it's super fast and bare-bones without all the metrics and graphs. Not a biggie and mostly unlikely, but worth a mention.

> deprecation strawman PR is up at #424

I will chip in, thanks for kicking it off 👍

**Comment by gerhard** [comment]
> @nesc58: How can I configure return_per_object_metrics configurable? For me it's a need to configure it.

For Docker Compose, I would recommend mounting the config file at `/etc/rabbitmq/conf.d/prometheus.return_per_object_metrics.conf`. The contents of the file would be:

```
prometheus.return_per_object_metrics = true
```

For Docker Swarm, I would recommend using a [service config](https://docs.docker.com/engine/swarm/configs/) to generate that config file.

For K8S, I would recommend a [ConfigMap](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/) to generate that config file.

**Comment by nesc58** [comment]
> For Docker Compose, I would recommend mounting the config file at `/etc/rabbitmq/conf.d/prometheus.return_per_object_metrics.conf`. The contents of the file would be:

Thank you. That was what I am looking for :-)

**Comment by michaelklishin** [comment]
What's blocking this from being merged @tianon @gerhard?

**Comment by tianon** [comment]
I think just your eyeballs to ensure my final (simplified) version is acceptable as a default. :sweat_smile:

File: update.sh
Comment (gerhard) [inline update.sh:151]: Now that https://github.com/docker-library/rabbitmq/pull/431 is merged, if we revert all these changes to `update.sh`, we can rebase on top of the current master, merge & ship it.

**Review by gerhard — CHANGES_REQUESTED** [review]
Almost there, will keep an eye on this PR until it makes it through 👀

File: update.sh
Comment (tianon) [reply update.sh]: Nice catch, updated! :+1:

**Comment by gerhard** [comment]
`rabbitmq:3.8` & `rabbitmq:3.8-management` image variants work as expected, OK to merge 👍

**Comment by jwillmer** [comment]
@tianon  I'm using the script below as entrypoint in my docker-compose file via: `entrypoint: ["bash", "/scripts/cluster-entrypoint.sh" ]`. How would I need to change it in order to get the same functionality?

```bash
#!/bin/bash

set -e

# Start RMQ from entry point.
# This will ensure that environment variables passed
# will be honored
/usr/local/bin/docker-entrypoint.sh rabbitmq-server -detached

# Do the cluster dance
rabbitmqctl stop_app
rabbitmqctl join_cluster rabbit@rabbitmq1

# Stop the entire RMQ server. This is done so that we
# can attach to it again, but without the -detached flag
# making it run in the forground
rabbitmqctl stop

# Wait a while for the app to really stop
sleep 2s

# Start it
rabbitmq-server
```

**Pull request merged**
